File: clients/ios/VisualPose/AnalysisJson.swift
import Foundation

enum AnalysisJson {
    static func save(_ file: URL, clipId: String, frames: [PoseFrame]) {
        var arr: [[String: Any]] = []
        for frame in frames {
            let poses: [[String: Any]] = frame.poses.map { pose in
                [
                    "keypoints": pose.keypoints.map { kp -> [String: Any] in
                        [
                            "x": Double(kp.x),
                            "y": Double(kp.y),
                            "confidence": Double(kp.confidence),
                        ]
                    },
                ]
            }
            let blaze: [[String: Any]] = frame.blaze33.map { joint in
                [
                    "x": Double(joint.x),
                    "y": Double(joint.y),
                    "z": Double(joint.z),
                    "confidence": Double(joint.confidence),
                ]
            }
            var obj: [String: Any] = [
                "t_ms": frame.tMs,
                "width": frame.width,
                "height": frame.height,
                "poses": poses,
                "blaze33": blaze,
            ]
            if let box = frame.bbox {
                obj["bbox"] = [Double(box.x1), Double(box.y1), Double(box.x2), Double(box.y2)]
            } else {
                obj["bbox"] = NSNull()
            }
            arr.append(obj)
        }
        Json.write(file, ["clip_id": clipId, "frames": arr], pretty: false)
    }

    static func load(_ file: URL) -> [PoseFrame] {
        guard Json.isFile(file),
              let text = try? String(contentsOf: file, encoding: .utf8),
              let root = Json.object(from: text),
              let arr = Json.optArray(root, "frames")
        else { return [] }
        var frames: [PoseFrame] = []
        for item in arr {
            guard let obj = item as? [String: Any] else { continue }
            var poses: [CocoPose] = []
            for poseItem in Json.optArray(obj, "poses") ?? [] {
                guard let poseObj = poseItem as? [String: Any],
                      let kps = Json.optArray(poseObj, "keypoints")
                else { continue }
                var points: [CocoKeypoint] = []
                for kpItem in kps {
                    guard let kp = kpItem as? [String: Any] else { continue }
                    points.append(
                        CocoKeypoint(
                            x: Float(Json.optDouble(kp, "x", 0)),
                            y: Float(Json.optDouble(kp, "y", 0)),
                            confidence: Float(Json.optDouble(kp, "confidence", 0))
                        )
                    )
                }
                if !points.isEmpty {
                    poses.append(CocoPose(keypoints: points))
                }
            }
            var blaze: [BlazeJoint] = []
            for jointItem in Json.optArray(obj, "blaze33") ?? [] {
                guard let joint = jointItem as? [String: Any] else { continue }
                blaze.append(
                    BlazeJoint(
                        x: Float(Json.optDouble(joint, "x", 0)),
                        y: Float(Json.optDouble(joint, "y", 0)),
                        z: Float(Json.optDouble(joint, "z", 0)),
                        confidence: Float(Json.optDouble(joint, "confidence", 0))
                    )
                )
            }
            let bbox: BBox?
            if let boxArr = Json.optArray(obj, "bbox"), boxArr.count >= 4,
               let x1 = Json.number(boxArr[0]),
               let y1 = Json.number(boxArr[1]),
               let x2 = Json.number(boxArr[2]),
               let y2 = Json.number(boxArr[3])
            {
                bbox = BBox(x1: Float(x1), y1: Float(y1), x2: Float(x2), y2: Float(y2))
            } else {
                bbox = nil
            }
            frames.append(
                PoseFrame(
                    tMs: Json.optInt64(obj, "t_ms"),
                    poses: poses,
                    width: max(Json.optInt(obj, "width", 1), 1),
                    height: max(Json.optInt(obj, "height", 1), 1),
                    blaze33: blaze,
                    bbox: bbox
                )
            )
        }
        return frames
    }

    static func toAssessFrames(_ frames: [PoseFrame]) -> [AssessFrame] {
        frames.map { AssessFrame(tMs: Double($0.tMs), blaze33: $0.blaze33) }
    }
}

File: clients/android/app/src/main/java/local/visual/corepose/AnalysisJson.kt
package local.visual.corepose

import org.json.JSONArray
import org.json.JSONObject
import java.io.File

object AnalysisJson {
    fun save(file: File, clipId: String, frames: List<PoseFrame>) {
        file.parentFile?.mkdirs()
        val arr = JSONArray()
        for (frame in frames) {
            val poses = JSONArray()
            for (pose in frame.poses) {
                val kps = JSONArray()
                for (kp in pose.keypoints) {
                    kps.put(
                        JSONObject()
                            .put("x", kp.x.toDouble())
                            .put("y", kp.y.toDouble())
                            .put("confidence", kp.confidence.toDouble()),
                    )
                }
                poses.put(JSONObject().put("keypoints", kps))
            }
            val blaze = JSONArray()
            for (joint in frame.blaze33) {
                blaze.put(
                    JSONObject()
                        .put("x", joint.x.toDouble())
                        .put("y", joint.y.toDouble())
                        .put("z", joint.z.toDouble())
                        .put("confidence", joint.confidence.toDouble()),
                )
            }
            val obj = JSONObject()
                .put("t_ms", frame.tMs)
                .put("width", frame.width)
                .put("height", frame.height)
                .put("poses", poses)
                .put("blaze33", blaze)
            if (frame.bbox == null) {
                obj.put("bbox", JSONObject.NULL)
            } else {
                obj.put(
                    "bbox",
                    JSONArray()
                        .put(frame.bbox.x1.toDouble())
                        .put(frame.bbox.y1.toDouble())
                        .put(frame.bbox.x2.toDouble())
                        .put(frame.bbox.y2.toDouble()),
                )
            }
            arr.put(obj)
        }
        file.writeText(
            JSONObject()
                .put("clip_id", clipId)
                .put("frames", arr)
                .toString(),
            Charsets.UTF_8,
        )
    }

    fun load(file: File): List<PoseFrame> {
        if (!file.isFile) {
            return emptyList()
        }
        val root = JSONObject(file.readText(Charsets.UTF_8))
        val arr = root.optJSONArray("frames") ?: return emptyList()
        val frames = ArrayList<PoseFrame>(arr.length())
        for (i in 0 until arr.length()) {
            val obj = arr.optJSONObject(i) ?: continue
            val posesArr = obj.optJSONArray("poses") ?: JSONArray()
            val poses = ArrayList<CocoPose>(posesArr.length())
            for (p in 0 until posesArr.length()) {
                val poseObj = posesArr.optJSONObject(p) ?: continue
                val kps = poseObj.optJSONArray("keypoints") ?: continue
                val points = ArrayList<CocoKeypoint>(kps.length())
                for (k in 0 until kps.length()) {
                    val kp = kps.optJSONObject(k) ?: continue
                    points.add(
                        CocoKeypoint(
                            x = kp.optDouble("x").toFloat(),
                            y = kp.optDouble("y").toFloat(),
                            confidence = kp.optDouble("confidence").toFloat(),
                        ),
                    )
                }
                if (points.isNotEmpty()) {
                    poses.add(CocoPose(points))
                }
            }
            val blazeArr = obj.optJSONArray("blaze33") ?: JSONArray()
            val blaze = ArrayList<BlazeJoint>(blazeArr.length())
            for (j in 0 until blazeArr.length()) {
                val joint = blazeArr.optJSONObject(j) ?: continue
                blaze.add(
                    BlazeJoint(
                        x = joint.optDouble("x").toFloat(),
                        y = joint.optDouble("y").toFloat(),
                        z = joint.optDouble("z").toFloat(),
                        confidence = joint.optDouble("confidence").toFloat(),
                    ),
                )
            }
            val boxArr = obj.optJSONArray("bbox")
            val bbox = if (boxArr != null && boxArr.length() >= 4) {
                BBox(
                    boxArr.optDouble(0).toFloat(),
                    boxArr.optDouble(1).toFloat(),
                    boxArr.optDouble(2).toFloat(),
                    boxArr.optDouble(3).toFloat(),
                )
            } else {
                null
            }
            frames.add(
                PoseFrame(
                    tMs = obj.optLong("t_ms"),
                    poses = poses,
                    width = obj.optInt("width", 1).coerceAtLeast(1),
                    height = obj.optInt("height", 1).coerceAtLeast(1),
                    blaze33 = blaze,
                    bbox = bbox,
                ),
            )
        }
        return frames
    }

    fun toAssessFrames(frames: List<PoseFrame>): List<AssessFrame> {
        return frames.map { frame ->
            AssessFrame(tMs = frame.tMs.toDouble(), blaze33 = frame.blaze33)
        }
    }
}
